src/main/java/org/opensearch/flowframework/rest/AbstractSearchWorkflowAction.java

```diff
@@ -0,0 +1,115 @@
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.BytesRestResponse;
+import org.opensearch.rest.RestChannel;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.rest.RestResponse;
+import org.opensearch.rest.action.RestResponseListener;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opensearch.core.xcontent.ToXContent.EMPTY_PARAMS;
+import static org.opensearch.flowframework.common.FlowFrameworkFeatureEnabledSetting.FLOW_FRAMEWORK_ENABLED;
+import static org.opensearch.flowframework.util.RestHandlerUtils.getSourceContext;
+
+/**
+ * Abstract class to handle search request.
+ */
+public abstract class AbstractSearchWorkflowAction<T extends ToXContentObject> extends BaseRestHandler {
+
+    protected final List<String> urlPaths;
+    protected final String index;
+    protected final Class<T> clazz;
+    protected final ActionType<SearchResponse> actionType;
+    protected final FlowFrameworkFeatureEnabledSetting flowFrameworkFeatureEnabledSetting;
+
+    /**
+     * Instantiates a new AbstractSearchWorkflowAction
+     * @param urlPaths urlPaths to create routes
+     * @param index index the search should be done on
+     * @param clazz model class
+     * @param actionType from which action abstract class is called
+     * @param flowFrameworkFeatureEnabledSetting Whether this API is enabled
+     */
+    public AbstractSearchWorkflowAction(
+        List<String> urlPaths,
+        String index,
+        Class<T> clazz,
+        ActionType<SearchResponse> actionType,
+        FlowFrameworkFeatureEnabledSetting flowFrameworkFeatureEnabledSetting
+    ) {
+        this.urlPaths = urlPaths;
+        this.index = index;
+        this.clazz = clazz;
+        this.actionType = actionType;
+        this.flowFrameworkFeatureEnabledSetting = flowFrameworkFeatureEnabledSetting;
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        if (!flowFrameworkFeatureEnabledSetting.isFlowFrameworkEnabled()) {
+            FlowFrameworkException ffe = new FlowFrameworkException(
+                "This API is disabled. To enable it, update the setting [" + FLOW_FRAMEWORK_ENABLED.getKey() + "] to true.",
+                RestStatus.FORBIDDEN
+            );
+            return channel -> channel.sendResponse(
+                new BytesRestResponse(ffe.getRestStatus(), ffe.toXContent(channel.newErrorBuilder(), ToXContent.EMPTY_PARAMS))
+            );
+        }
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.parseXContent(request.contentOrSourceParamParser());
+        searchSourceBuilder.fetchSource(getSourceContext(request, searchSourceBuilder));
+        searchSourceBuilder.seqNoAndPrimaryTerm(true).version(true);
```

Comment:
I don't think we currently update seqNo and Primary term but its maybe something we should add. 
They help with concurency control when updating the index

Reply:
I am guessing this is something that will be handled with AccessController.